src/article.cc

```diff
@@ -173,21 +168,19 @@ Napi::Value Article::getData(const Napi::CallbackInfo& info) {
                                     article_->bufferData.size());
 }
 void Article::setData(const Napi::CallbackInfo& info, const Napi::Value& data) {
-  std::vector<uint8_t> buffer;
-  {
-    if (data.IsString()) {
-      std::string nbuf = data.ToString();
-      buffer.resize(nbuf.size());
-      std::memcpy(buffer.data(), nbuf.data(), nbuf.size());
-    } else if (data.IsBuffer()) {
-      auto nbuf = data.As<Napi::Buffer<uint8_t>>();
-      buffer.resize(nbuf.Length());
-      std::memcpy(buffer.data(), nbuf.Data(), nbuf.Length());
-    } else {
-      throw Napi::Error::New(info.Env(), "data must be a string or buffer");
-    }
+  if (data.IsString()) {
+    std::string nbuf = data.ToString();
+    article_->setData(nbuf.begin(), nbuf.end());
```

Comment:
I am not a N-API expert, so I may be wrong (or right but we cannot fix it easily).

With `std::string nbuf = data.ToString();` you are creating a string containing the data.
Then you pass the begin and end of the string to setData which in turn will copy the data (create a std::vector<uint8_t>) into bufferData.

So, you are creating a string (copying the data) just as temporal buffer. You are copying the data twice (from data to nbuf and from nbuf to bufferData) instead of once (from data to bufferData).

To fix this, I see two solutions (but N-API need to provide it):
- if `data.ToString()` return a reference to a string, keep the reference instead of copying the data into a new string. So `std::string& nbuf = data.ToString()`
- if it is possible to directly have access to the pointer of data (and its length), pass the pointer to `setData` directly (`article_->setData(data.pointer(), data.pointer()+data.size())`)

And this is probably the same problem when data is a buffer.

But as I said, I don't know if this is easily fixable, it depends of N-API and I don't know N-API.

Reply:
The [ToString](https://github.com/nodejs/node-addon-api/blob/master/napi.h#L302) is located [here](https://github.com/nodejs/node-addon-api/blob/master/napi.h#L302) and returns a [Napi::String](https://github.com/nodejs/node-addon-api/blob/master/napi.h#L418) which has a [operator std::string()](https://github.com/nodejs/node-addon-api/blob/master/napi.h#L471) method. I don't currently see a way to get direct access to the data.

Reply:
It seems it is not possible indeed. Let's close this point.